### PR TITLE
Tribal Section 4 Parsing & Validation

### LINKED ISSUE #1117: Tribal TANF Stratum Data (04) File Parsing - As a data analyst, I want my data files to be parsed so that my submission can be validated/accepted by TDP
**Description:**
_Provide a brief background and justification for this issue_
Upon file upload, we expect TDP to attempt parsing per the spec outlined:
https://www.acf.hhs.gov/sites/default/files/documents/ofa/tanf_data_report_section4.pdf

**Acceptance Criteria:** 
_Create a list of functional outcomes that must be achieved to complete this issue_

- [ ] Allow Tribes to submit Closed Data based on the TANF data parsing spec 
- [ ] Testing Checklist has been run and all tests pass
- [ ] README is updated, if necessary

**Tasks:**
_Create a list of granular, specific work items that must be completed to deliver the desired outcomes of this issue_ 
- [ ] Ensuring file content type is tribal TANF. 
- [ ] Correctly formatted input data can be built into a Python data structure
- [ ] Completed Python data structure can be translated to database model(s)
- [ ] Incorrectly formatted data is identified and mistake(s) are logged.
- [ ] Run Testing Checklist and confirm all tests pass

**Notes:**
_Add additional useful information, such as related issues and functionality that isn't covered by this specific issue, and other considerations that will be helpful for anyone reading this_ 

**Supporting Documentation:**
_Please include any relevant log snippets/files/screen shots_

**Open Questions:**
_Please include any questions or decisions that must be made before beginning work or to confidently call this issue complete_

### PR DESCRIPTION
**Pull request by elipe17**
## Summary of Changes
- Added support for section 4 parsing and validation
Pull request closes #1117 

## How to Test
```
cd tdrs-frontend && docker-compose up
cd tdrs-backend && docker-compose up
```

- Open http://localhost:3000/ and sign in.
- Submit `tribal_section_4_fake.txt`.
- Verify the file generates Tribal T7 records.

## Deliverables
_More details on how deliverables herein are assessed included [here](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
- [x] Allow Tribes to submit stratum Data based on the TANF data parsing spec 
- [x] Testing Checklist has been run and all tests pass
- [x] README is updated, if necessary

### [Deliverable 2: Tested Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [x] If this PR introduces backend code changes, are they meaningfully tested?
  + [x] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [x] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [x] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [x] Are backend code style checks passing on CircleCI?
+ [x] Are frontend code style checks passing on CircleCI?
+ [x] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?

### [Deliverable 5: Deployed](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [x] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [x] Does this PR provide background for why coding decisions were made?
+ [ ] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [x] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [x] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

### CODE REVIEW

**Review by raftmsohani — CHANGES_REQUESTED** [review]
please add friendly_name to t7 fields

**Review by jtimpe — APPROVED** [review]

**Review by raftmsohani — APPROVED** [review]
LGTM!

File: tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t7.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+                    item="3",
+                    name="CALENDAR_QUARTER",
+                    friendly_name="calendar quarter",
+                    type="number",
+                    startIndex=2,
+                    endIndex=7,
+                    required=True,
+                    validators=[
+                        validators.dateYearIsLargerThan(1998),
+                        validators.quarterIsValid(),
```
Comment (ADPennington) [inline tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t7.py:52]: @elipe17 i discovered an issue with this validator. 

- If the quarter is a space, then no error is raised in the logs or the feedback report
- if the quarter = 0 or > 4, then this message is returned in the logs `Invalid quarter value.` and the processing status stays in the "pending" state which does not generate a feedback report. 

I believe this is an edge case, but does need to be addressed in a new ticket.

Comment (ADPennington) [reply tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t7.py:52]: #2807

File: tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t2.py [hunk 2]
```diff
@@ -348,8 +348,8 @@
                 ),
                 Field(
                     item="39",
-                    name="PARENT_WITH_MINOR_CHILD",
-                    friendly_name="parent with minor child",
+                    name="PARENT_MINOR_CHILD",
```
Comment (ADPennington) [inline tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t2.py:351]: checked this. lgtm 👍🏾

**Review by ADPennington — APPROVED** [review]
@elipe17 this is ready to merge. i created a new ticket to clean up the validation for cal qtr since this field is present in several schemas. Great work! 🚀

**Pull request merged**